scipy/signal/_spectral_py.py

```diff
@@ -86,70 +116,186 @@ def lombscargle(x,
+    >>> fig, (ax_t, ax_p, ax_n, ax_a) = plt.subplots(4, 1, figsize=(5, 5))
     >>> ax_t.plot(x, y, 'b+')
     >>> ax_t.set_xlabel('Time [s]')
+    >>> ax_t.set_ylabel('Amplitude')
+
+    Then plot the periodogram for each of the normalize options:
+
+    >>> ax_p.plot(w, pgram_power)
+    >>> ax_p.set_xlabel('Angular frequency [rad/s]')
+    >>> ax_p.set_ylabel('Power')
 
-    Then plot the normalized periodogram:
+    >>> ax_n.plot(w, pgram_norm)
+    >>> ax_n.set_xlabel('Angular frequency [rad/s]')
+    >>> ax_n.set_ylabel('Normalized')
 
-    >>> ax_w.plot(w, pgram)
-    >>> ax_w.set_xlabel('Angular frequency [rad/s]')
-    >>> ax_w.set_ylabel('Normalized amplitude')
+    >>> ax_a.plot(w, np.abs(pgram_amp))
+    >>> ax_a.set_xlabel('Angular frequency [rad/s]')
+    >>> ax_a.set_ylabel('Amplitude')
+
+    >>> plt.tight_layout()
     >>> plt.show()
 
     """
-    x = np.ascontiguousarray(x, dtype=np.float64)
-    y = np.ascontiguousarray(y, dtype=np.float64)
-    freqs = np.ascontiguousarray(freqs, dtype=np.float64)
 
-    assert x.ndim == 1
-    assert y.ndim == 1
-    assert freqs.ndim == 1
+    # if no weights are provided, assume all data points are equally important
+    if weights is None:
+        weights = np.ones_like(x, dtype=x.dtype)
+
+    # validate input shapes
+    if not (x.ndim == 1 and x.size > 0 and x.shape == y.shape == weights.shape):
+        raise ValueError("Parameters x, y, weights must be 1-D arrays of "
+                         "equal non-zero length!")
+    if not (freqs.ndim == 1 and freqs.size > 0):
+        raise ValueError("Parameter freqs must be a 1-D array of non-zero length!")
+
+    # validate input dtypes
+    if not np.issubdtype(x.dtype, np.floating) or (
+        not (x.dtype == y.dtype == freqs.dtype == weights.dtype)):
+        raise TypeError("Parameters x, y, freqs, and weights must be arrays of same "
+                         "floating dtype (int and complex dtypes are not allowed)!")
+
+    # validate weights
+    if not (np.all(weights >= 0) and np.sum(weights) > 0):
+        raise ValueError("Parameter weights must have only non-negative entries "
+                         "which sum to a positive value!")
+
+    # validate normalize parameter
+    if isinstance(normalize, bool):
+        # if bool, convert to str literal
+        normalize = "normalize" if normalize else "power"
+
+    if normalize not in ["power", "normalize", "amplitude"]:
+        raise ValueError(
+            "Normalize must be: False (or 'power'), True (or 'normalize'), "
+            "or 'amplitude'."
+        )
+
+    # weight vector must sum to 1
+    weights = weights / weights.sum()
+
+    # pre-allocate arrays for intermediate and final calculations
+    coswt = np.empty_like(x)
+    sinwt = np.empty_like(x)
+    pgram = np.empty_like(freqs)
+    # store a and b so that phase can be calculated outside loop, if necessary
+    a = np.empty_like(freqs)
+    b = np.empty_like(freqs)
```

Comment:
@adammj with regard to checking the dtypes are all the same, if we did something like the following would the check be necessary or am I missing something?

```suggestion
    dtype = np.results_type(x, y, freqs)
    coswt = np.empty_like(x, dtype=dtype)
    sinwt = np.empty_like(x, dtype=dtype)
    pgram = np.empty_like(freqs, dtype=dtype)
    # store a and b so that phase can be calculated outside loop, if necessary
    a = np.empty_like(freqs, dtype=dtype)
    b = np.empty_like(freqs, dtype=dtype)
```

Reply:
So, we'll need to keep the` np.issubdtype(x.dtype, np.floating)` regardless, because the math is not valid for complex values (or integer, for that matter). This was also true for the old function, but it was never checked. Numpy would've thrown a warning if the user used complex values when the values were converted to np.float64. But neither numpy nor the function threw an error. If we allowed any mixed dtypes, we'll need to duplicate that check for y, freqs, and weights.

I guess we could use the `np.result_type` to get the promoted type and avoid the check that all of the types are the same (`x.dtype == y.dtype == freqs.dtype == weights.dtype`). I just wonder about the performance impact of mixing data types. Also, I guess I would then change the declaration of `weights = np.ones_like(x, dtype=x.dtype)` to `weights = np.ones_like(x, dtype=y.dtype)`, to more closely match the what is being weighted (y versus x). Furthermore, your `dtype = np.results_type(x, y, freqs)` would have to change to `dtype = np.results_type(x, y, freqs, weights)` to account for the case that the user provided weights in some other type.

Reply:
Looking more carefully, I see that casting to float64 was a document feature of the old code. I think my preference would just be to maintain this behaviour. That way, everything is good backwards compatibility-wise. What do you think? Sorry for dragging this out, this should hopefully be the last detail to sort out before merging!

Reply:
I'm totally fine with casting to float64 (I was doing that on the first commit). But, as DietBru said above, it might make sense to give the user the option to choose their own type.

> I think it is reasonable to convert the input to np.float64, but it should be documented. I personally prefer raising an error over converting types, since it opens up potential optimizations in the calling code.
> 
> There might be a niche use case, where a  np.float32 or np.float16 would be preferred: Evaluating if it makes sense to utilize a smaller float type for an optimized implementation (e.g., for a C implementation on an embedded device).

Reply:
I think I my preference would be cast to float64 for simplicity (and ease of verifying backwards compatibility) then if a user asks for other dtypes we can add it using the approach you have suggested in the future.

Reply:
Done. I also had to eliminate one of my added tests, as it was no longer relevant.

Reply:
Thanks, I appreciate the patience here. Trying to modify existing code is always a bit frustrating